File: sdk/python/scripts/extract_surface.py
#!/usr/bin/env python3
"""Emit the Python SDK's public surface as JSON, for the cross-SDK parity gate.

Reads the hand-written ergonomic layer with ``ast`` rather than importing it, so
this runs with no dependencies and no side effects, and reports what the source
actually declares rather than what an import happens to expose.

The generated typed core is deliberately out of scope: ``sdk-drift`` already
holds that to ``openapi/openapi.json``. What had no gate at all was the
hand-written layer, which is how the TypeScript client fell a release behind the
Python one without anything going red.

Usage:  python scripts/extract_surface.py > surface.python.json
"""

from __future__ import annotations

import ast
import json
import sys
from pathlib import Path
from typing import Any, Dict, List

PACKAGE = Path(__file__).resolve().parents[1] / "constitutional_aiops"

# Vocabulary constants that both clients mirror from the server. Compared by
# value, not just by name: a new insight kind landing in one client and not the
# other is exactly the drift this gate exists to catch.
TRACKED_CONSTANTS = (
    "INSIGHT_KINDS",
    "INSIGHT_TIERS",
    "INSIGHT_UNAVAILABLE_REASONS",
    "CONSTITUTIONAL_CODES",
)


def _module(name: str) -> ast.Module:
    path = PACKAGE / name
    return ast.parse(path.read_text(encoding="utf-8"), filename=str(path))


def _public_methods(module: ast.Module, class_name: str) -> List[str]:
    for node in module.body:
        if isinstance(node, ast.ClassDef) and node.name == class_name:
            return sorted(
                item.name
                for item in node.body
                if isinstance(item, (ast.FunctionDef, ast.AsyncFunctionDef))
                and not item.name.startswith("_")
            )
    raise SystemExit(f"parity: class {class_name} not found in the Python client")


def _error_classes(module: ast.Module) -> List[str]:
    return sorted(node.name for node in module.body if isinstance(node, ast.ClassDef))


def _constants(*modules: ast.Module) -> Dict[str, Any]:
    found: Dict[str, Any] = {}
    for module in modules:
        for node in module.body:
            if not isinstance(node, ast.Assign):
                continue
            for target in node.targets:
                if isinstance(target, ast.Name) and target.id in TRACKED_CONSTANTS:
                    try:
                        found[target.id] = list(ast.literal_eval(node.value))
                    except ValueError:
                        # A constant that is not a literal sequence is a change
                        # worth failing on rather than silently skipping.
                        raise SystemExit(
                            f"parity: {target.id} is not a literal sequence in the Python client"
                        )
    return found


def main() -> None:
    client = _module("client.py")
    errors = _module("errors.py")

    surface = {
        "language": "python",
        "methods": _public_methods(client, "AIOpsClient"),
        "errors": _error_classes(errors),
        "constants": {key: _constants(client, errors).get(key) for key in TRACKED_CONSTANTS},
    }
    json.dump(surface, sys.stdout, indent=2, sort_keys=True)
    sys.stdout.write("\n")


if __name__ == "__main__":
    main()

File: sdk/typescript/scripts/extract-surface.mjs
#!/usr/bin/env node
/**
 * Emit the TypeScript SDK's public surface as JSON, for the cross-SDK parity gate.
 *
 * Uses the TypeScript compiler's own parser rather than matching on text, so a
 * reformat, a comment or a multi-line signature cannot change the answer. Lives
 * inside sdk/typescript so `typescript` resolves from this package's own
 * node_modules after `npm ci`.
 *
 * Usage:  node scripts/extract-surface.mjs > surface.typescript.json
 */

import { readFileSync } from 'node:fs'
import { dirname, resolve } from 'node:path'
import { fileURLToPath } from 'node:url'
import ts from 'typescript'

const SRC = resolve(dirname(fileURLToPath(import.meta.url)), '..', 'src')

// Mirrored from the server, and compared by value in the parity check: a new
// insight kind reaching one client and not the other is the drift this catches.
const TRACKED_CONSTANTS = [
  'INSIGHT_KINDS',
  'INSIGHT_TIERS',
  'INSIGHT_UNAVAILABLE_REASONS',
  'CONSTITUTIONAL_CODES',
]

function parse(name) {
  const path = resolve(SRC, name)
  return ts.createSourceFile(path, readFileSync(path, 'utf8'), ts.ScriptTarget.Latest, true)
}

function isExported(node) {
  return (ts.getCombinedModifierFlags(node) & ts.ModifierFlags.Export) !== 0
}

function isPrivate(node) {
  return (ts.getCombinedModifierFlags(node) & ts.ModifierFlags.Private) !== 0
}

/** Public method names on a class, excluding the constructor and private members. */
function publicMethods(source, className) {
  let methods = null
  source.forEachChild((node) => {
    if (ts.isClassDeclaration(node) && node.name?.text === className) {
      methods = node.members
        .filter((m) => ts.isMethodDeclaration(m) && !isPrivate(m) && m.name)
        .map((m) => m.name.getText(source))
        .sort()
    }
  })
  if (methods === null) {
    console.error(`parity: class ${className} not found in the TypeScript client`)
    process.exit(1)
  }
  return methods
}

/** Exported class names, which is how the error hierarchy is declared. */
function exportedClasses(source) {
  const names = []
  source.forEachChild((node) => {
    if (ts.isClassDeclaration(node) && node.name && isExported(node)) names.push(node.name.text)
  })
  return names.sort()
}

/** Exported `const NAME = [...] as const` string arrays, read as real values. */
function constants(...sources) {
  const found = {}
  for (const source of sources) {
    source.forEachChild((node) => {
      if (!ts.isVariableStatement(node) || !isExported(node)) return
      for (const decl of node.declarationList.declarations) {
        const name = decl.name.getText(source)
        if (!TRACKED_CONSTANTS.includes(name) || !decl.initializer) continue
        // `[...] as const` wraps the array in an assertion expression.
        const init = ts.isAsExpression(decl.initializer)
          ? decl.initializer.expression
          : decl.initializer
        if (!ts.isArrayLiteralExpression(init)) {
          console.error(`parity: ${name} is not an array literal in the TypeScript client`)
          process.exit(1)
        }
        found[name] = init.elements.map((el) => {
          if (!ts.isStringLiteral(el)) {
            console.error(`parity: ${name} holds a non-string entry in the TypeScript client`)
            process.exit(1)
          }
          return el.text
        })
      }
    })
  }
  return found
}

const client = parse('client.ts')
const errors = parse('errors.ts')
const found = constants(client, errors)

const surface = {
  language: 'typescript',
  methods: publicMethods(client, 'AIOpsClient'),
  errors: exportedClasses(errors),
  constants: Object.fromEntries(TRACKED_CONSTANTS.map((key) => [key, found[key] ?? null])),
}

process.stdout.write(`${JSON.stringify(surface, null, 2)}\n`)
